base/multidimensional.jl

```diff
@@ -440,44 +480,69 @@ module IteratorsMD
             return nothing
         end
         iterfirst, iterfirst
     end
     @inline function iterate(r::Reverse{<:CartesianIndices}, state)
-        valid, I = __dec(state.I, last(r.itr).I, first(r.itr).I)
+        valid, I = __dec(state.I, r.itr.indices)
         valid || return nothing
         return CartesianIndex(I...), CartesianIndex(I...)
     end
 
     # decrement & carry
-    @inline function dec(state, start, stop)
-        _, I = __dec(state, start, stop)
+    @inline function dec(state, indices)
+        _, I = __dec(state, indices)
         return CartesianIndex(I...)
     end
 
     # decrement post check to avoid integer overflow
-    @inline __dec(::Tuple{}, ::Tuple{}, ::Tuple{}) = false, ()
-    @inline function __dec(state::Tuple{Int}, start::Tuple{Int}, stop::Tuple{Int})
-        valid = state[1] > stop[1]
-        return valid, (state[1]-1,)
+    @inline __dec(::Tuple{}, ::Tuple{}) = false, ()
+    @inline function __dec(state::Tuple{Int}, indices::Tuple{<:OrdinalRange})
+        rng = indices[1]
+        I = state[1] - step(rng)
+        valid = __is_valid_range(I, rng) && state[1] != first(rng)
+        return valid, (I,)
     end
 
-    @inline function __dec(state, start, stop)
-        if state[1] > stop[1]
-            return true, (state[1]-1, tail(state)...)
+    @inline function __dec(state, indices)
+        rng = indices[1]
+        I = state[1] - step(rng)
+        if __is_valid_range(I, rng) && state[1] != first(rng)
+            return true, (I, tail(state)...)
         end
-        valid, I = __dec(tail(state), tail(start), tail(stop))
-        return valid, (start[1], I...)
+        valid, I = __dec(tail(state), tail(indices))
+        return valid, (last(rng), I...)
     end
 
     # 0-d cartesian ranges are special-cased to iterate once and only once
     iterate(iter::Reverse{<:CartesianIndices{0}}, state=false) = state ? nothing : (CartesianIndex(), true)
 
-    Base.LinearIndices(inds::CartesianIndices{N,R}) where {N,R} = LinearIndices{N,R}(inds.indices)
+    Base.LinearIndices(inds::CartesianIndices{N,R}) where {N,R<:AbstractUnitRange} = LinearIndices{N,R}(inds.indices)
+    function Base.LinearIndices(inds::CartesianIndices{N}) where N
+        indices = inds.indices
+        if all(x->x==1, step.(indices))
+            indices = map(ind->convert(AbstractUnitRange, ind), indices)
+            LinearIndices{N, typeof(indices)}(indices)
+        else
+            throw(ArgumentError("LinearIndices for $typeof(inds) with non-1 step size is not yet supported."))
+        end
+    end
+
+    # This is needed because converting to LinearIndices is only available when steps are all 1
+    # NOTE: this is only a temporary patch until the LinearIndices support is done
+    function Base.collect(inds::CartesianIndices)
+        dest = Array{eltype(inds)}(undef, size(inds))
+        @inbounds for (i, idx) in enumerate(inds)
+            dest[i] = idx
+        end
+        dest
+    end
```

Comment:
I'm a bit worried about this:

```julia
julia> @btime collect(CartesianIndices((64, 64)));
  5.954 μs (2 allocations: 64.08 KiB) # this
  4.044 μs (2 allocations: 64.08 KiB) # master
```

We should fix this before this gets merged.